Messaging function

robot269006 opened the pull request:
# WHAT
Configure message creating function

# WHY
So that the users can send messages in their group

robot269006 committed:
Create messages model & table

1. rails g model Messages
2. Edit migration file (no image column at this point)
3. rake db:migrate
4. Rename user_id and group_id (correction)
5. Add index to messages table (user_id and group_id)
6. rake db:migrate

robot269006 committed:
Set association

1. Set association on message.rb
2. Set association on group.rb and user.rb

robot269006 committed:
Rename top_controller -> messages_controller

1. rails g controller messages
2. Move index.html.haml to views/messages
3. rails g destroy controller top
4. Re-route -> resources messages only: [:index, :create]
5. Re-route rooting -> root ‘messages#index’

robot269006 committed:
Set controller

1. Send messages instance to index method
2. Send message instance to create method

robot269006 committed:
Set view files

1. Cut out the individual messages from index
2. Make new file “_messages.html.haml”
3. Set render in index
4. Edit partial “_messages” ->
*@messages instance deleted temporarily from def index, so it can be
tested

robot269006 committed:
Set instances

1. Try handing over @user instances first
2. Try handing over @group instances

*Error in @group = Group.find(params[:group_id])
-> it works if “find(3)” is put in. So it must be something with
association?

robot269006 committed:
Fix root path

root path deleted -> it was set to ‘messages#index’, but this doesn’t
lead to index page since group_id parameter can’t be sent. The group_id
parameter is not sent through the path settings.

So temporarily altered the root path to sessions#new

robot269006 committed:
Revert "Fix root path"

This reverts commit 38f444e249e8be72926194417a3a362f63bd3cf6.

robot269006 committed:
Fix rooting

robot269006 committed:
Fix root path

1. It was written as :groups_id -> :group_id
2. Set root path to something that doesn’t require any id-dependent info

robot269006 committed:
Flash message

Added flash messages

robot269006 committed:
View fix (redirect_to routing, link settings)

1. Fix flash messages and redirect_to routing
2. Set links to icons
3. Delete rendering files (complicates later view settings, so delete
for now)

robot269006 committed:
Enable message creation

1. Fix minor mistake on view (indent on name, date, body)
2. Create @message instance on index
3. Edit index -> form_for (pass on @group and @message)
4. Fix strong parameters (message_params)

robot269006 committed:
Edit view (Left content)

1. Add user to several groups -> so it can be tested later
2. Create @groups instance -> has all the user-related groups
3. Edit view, @groups.each do

*Check if scss is applied correctly
*Fix message part

robot269006 committed:
Edit view (Right-bottom content)

1. Minor description change -> user name display from current_user.name
to @user.name, each |group| to each |g|

2. Set @group.messages (check with binding.pry) -> separate with .each
|m| method

robot269006 committed:
Deal with 'Fixme'

1. Fixme on left-content -> g.messages.all.last.body
2. Fixme on right-bottom content -> m.user.name
3. Set link to switch groups on left-content -> link_to
group_messages_path
4. Group member display -> @group.user.each -> user.name

robot269006 committed:
Delete @user instance

@user instance replaced with current_user

robot269006 committed:
Set validation

1. Set validation to message.rb
2. Set render to :index
3. Set @groups to when the message creation fails

robot269006 commented:
<img width="1266" alt="2017-05-01 9 15 08" src="https://cloud.githubusercontent.com/assets/24761206/25569257/bbd2ace6-2e4e-11e7-950f-a9df612bdfa2.png">

ta910 commented on app/controllers/messages_controller.rb:
```diff
@@ -0,0 +1,30 @@
+      redirect_to group_messages_path(@group), notice: "メッセージが作成されました"
+    else
+      @groups = current_user.groups
+      flash[:alert]="メッセージの作成に失敗しました"
```
flash.nowを用いましょう。参考にしてください。
http://trwtnb.blogspot.jp/2009/11/rubyrailsflashnownoticeflashnotice.html

robot269006 replied:
ありがとうございます！全然知らなかったので参考になりました。

ta910 commented on app/controllers/messages_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  private
+  def message_params
+    params.require(:message).permit(:body).merge(group_id: params[:group_id], user_id: current_user.id)
+  end
```
この後に一行空行を開けましょう。

robot269006 replied:
ありがとうございます、対応させていただきました。

ta910 commented on app/controllers/messages_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  def message_params
+    params.require(:message).permit(:body).merge(group_id: params[:group_id], user_id: current_user.id)
+  end
+  def group_params
```
グループを探して持って来ているので動詞を使って、set_groupなどにメソッド名を変更しましょう。

robot269006 replied:
こちらもありがとうございます、対応させていただきました。

ta910 commented on app/views/messages/index.html.haml:
```diff
@@ -0,0 +1,31 @@
+        %ul.message
+        = link_to group_messages_path(g) do
+          %p.message__group= g.name
+          %p.message__body= g.messages.all.last.body
```
メッセージがまだないときの分岐を入れましょう。

robot269006 replied:
ご指摘ありがとうございます、対応させていただきました。

ta910 commented on app/views/messages/index.html.haml:
```diff
@@ -0,0 +1,31 @@
+    .group__top
+      .groupdata
+        %p.groupdata__name= @group.name
+        - @group.users.each do |n|
```
eachでなく部分テンプレートを使いましょう。

ta910 commented on app/views/messages/index.html.haml:
```diff
@@ -0,0 +1,31 @@
+        = link_to '', edit_user_registration_path, class: 'fa fa-cog icon__edit'
+        = link_to '', new_user_registration_path, class: 'fa fa-pencil-square-o icon__setting'
+    .navi__bottom
+      - @groups.each do |g|
```
eachでなく、部分テンプレートを使いましょう。

robot269006 replied:
ありがとうございます、修正させていただきました

ta910 commented on app/views/messages/index.html.haml:
```diff
@@ -0,0 +1,31 @@
+          %p.groupdata__memberlist= n.name
+      = link_to 'Edit', edit_group_path(@group), class: 'editbutton'
+    .group__middle
+      - @group.messages.each do |m|
```
こちらも部分テンプレートを用いましょう。

robot269006 replied:
ありがとうございます、修正しました

ta910 commented on app/views/messages/index.html.haml:
```diff
@@ -0,0 +1,31 @@
+      - @group.messages.each do |m|
+        %ul.individualmessage
+          %li.individualmessage__name= m.user.name
+          %li.individualmessage__date= m.created_at
```
時間を何年何月のように日本語表記されるようにしましょう。

robot269006 replied:
ありがとうございます、対応させていただきました！

ta910 commented on config/routes.rb:
```diff
@@ -1,5 +1,7 @@
-  root 'top#index'
-  resources 'groups'
+  root 'groups#new'
+  resources :groups do
```
使っているアクションだけに絞っておきましょう。

robot269006 replied:
ありがとうございます、修正しました

ta910 commented on config/routes.rb:
```diff
@@ -1,5 +1,7 @@
   devise_for :users
-  root 'top#index'
-  resources 'groups'
+  root 'groups#new'
```
rootはgroupのインデックスが望ましいです。
ログイン後、必ずグループを作らなければいけなくなってしまいます。

robot269006 replied:
ご指摘ありがとうございます、対応させていただきました

robot269006 committed:
Revision (Minor fixes)

1. flash[:alert] -> flash.now[:alert], so that it limits the flash
message to current request
2. Naming from group_params -> set_group, to include verb in the naming

robot269006 committed:
Revision (date)

used strftime method to convert from default to japanese

robot269006 committed:
Revision (render)

1. Create render files
2. Whack in the cut-pit parts

*Repeat above for navibottom and group middle parts
*Deal with taking out ‘each’ description later

robot269006 committed:
Revision (Routing)

1. Narrow down the resources only to the used actions
2. Create index action to groups_controller
3. Set root to group#index
4. Create group#index view file

KoichiOzaki commented on app/controllers/messages_controller.rb:
```diff
@@ -0,0 +1,31 @@
+    if @message.save
+      redirect_to group_messages_path(@group), notice: "メッセージが作成されました"
+    else
+      @groups = current_user.groups
```
この変数定義は必要でしょうか？

robot269006 replied:
いえ、どちらにせよindexをrenderするので必要ないと思います。削除します！

KoichiOzaki commented on app/views/messages/_groupmiddle.html.haml:
```diff
@@ -0,0 +1,5 @@
+  %ul.individualmessage
+    %li.individualmessage__name= m.user.name
+    %li.individualmessage__date= m.created_at.strftime("%Y年%m月%d日 %H時%M分")
+    %li.individualmessage__body= m.body
```
部分テンプレートとしては
2行目以降の部分を
_message.html.hamlとして定義したいですね。

robot269006 replied:
対応させていただきました。ご指摘ありがとうございます！

KoichiOzaki commented on app/views/messages/_navibottom.html.haml:
```diff
@@ -0,0 +1,8 @@
+- @groups.each do |g|
```
こちらも
2行目以降の部分を
_group.html.hamlとして定義しましょう

robot269006 replied:
ご指摘ありがとうございます、対応しました！

robot269006 committed:
Revision (each -> render)

1. Delete “each” description from rendered files
2. Create @messages instance in order to pass message info
3. Allow rendering short-cut with “render @groups” and “render
@messages”
4. Rename rendered files to _group.html.haml and _message.html.haml

robot269006 committed:
Revision (delete unnecessary @groups)

delete @groups from message creation failure
(already passed on since it renders index)
